sympy/ntheory/qs.py

```diff
@@ -0,0 +1,234 @@
+from sympy import isprime, log, sqrt, exp, sieve, gcd, mod_inverse, ceiling
+from sympy.ntheory.residue_ntheory import _sqrt_mod_prime_power
+
+class Polynomial:
+    def __init__(self, modified_coeff=[], a=None, b=None):
+        """ Here a and b are the parameters of the polynomial, and
+        modified_coeff is modified coeff to reduce amount of
+        calculation whenever eval is called.
+        modified_coeff = [a**2, 2*a*b, b**2 - N]"""
+        self.modified_coeff = modified_coeff
+        self.a = a
+        self.b = b
+
+    def eval(self, x):
+        ans = 0
+        for coeff in self.modified_coeff:
+            ans *= x
+            ans += coeff
+        return ans
+
+class factor_base_elem:
+    def __init__(self, prime, tmem_p, log_p):
+        self.prime = prime
+        self.tmem_p = tmem_p
+        self.log_p = log_p
+        self.soln1 = None
+        self.soln2 = None
+        self.a_inv = None
+        self.B_ainv = None
+
+
+
+def gen_factor_base(B, n):
+    factor_base = [2, 3]
+    for p in sieve.primerange(4, B + 1):
+        a = n % p
+        if a < 2 or pow(a, (p - 1) // 2, p) == 1:
+            residue = _sqrt_mod_prime_power(n, p, 1)
+            log_p = round(log(p, 2).evalf())
+            factor_base.append(factor_base_elem(p, residue, log_p))
+    return factor_base
+
+
+def sieve_poly(N, M, factor_base):
+    """Initialization stage for first polynomial"""
+    import random
+    approx_val = sqrt(2*N).evalf() / M
+    #randomly search for a combination of primes whose multiplication is close to approx_val
+    best_a, best_q, best_ratio = None, None, None
+    for _ in range(50):
```

Comment:
Here we try to  get a  subset of factor base such that the multiplication of this subset is  close to `approx_value`. Here  i am using  a random method to get it but the ratio comes around ~0.3. Is tit possible to use dynamic programming to get a better approximation for `a`.